### ISSUE TITLE
Disabling then enabling operator in the Helm chart causes all resources to be deleted.

### ISSUE THREAD

**Issue by EinarArnason** [gravitational/teleport#31059]
Expected behavior:
Resources to be still in place after upgrade

Current behavior:
No resources are present. teleport-agent pods fail with `Readiness probe failed: HTTP probe failed with statuscode: 400`

Bug details:
- Teleport version 13.3.6
- Recreation steps helm upgrade teleport teleport/teleport-cluster --reuse-values
- Debug logs 
teleport-auth pod:

```
2023-08-27T09:09:52Z INFO [UPLOAD:2]  Creating directory /var/lib/teleport/log/upload/streaming/default. pid:6.2 service/service.go:2744
2023-08-27T09:09:52Z INFO [UPLOAD:2]  Creating directory /var/lib/teleport/log. pid:6.2 service/service.go:2744
2023-08-27T09:09:52Z WARN [PROC:1]    Teleport process has exited with error. error:[
ERROR REPORT:
Original Error: *errors.errorString context canceled
Stack Trace:
github.com/gravitational/teleport/lib/auth/server_info.go:56 github.com/gravitational/teleport/lib/auth.(*Server).ReconcileServerInfos
github.com/gravitational/teleport/lib/service/service.go:2013 github.com/gravitational/teleport/lib/service.(*TeleportProcess).initAuthService.func7
github.com/gravitational/teleport/lib/service/supervisor.go:546 github.com/gravitational/teleport/lib/service.(*LocalService).Serve
github.com/gravitational/teleport/lib/service/supervisor.go:281 github.com/gravitational/teleport/lib/service.(*LocalSupervisor).serve.func1
runtime/asm_amd64.s:1598 runtime.goexit
User Message: context canceled] pid:6.1 service:auth.server_info service/supervisor.go:286
2023-08-27T09:09:52Z INFO [UPLOAD:2]  Creating directory /var/lib/teleport/log/upload. pid:6.2 service/service.go:2744
2023-08-27T09:09:52Z WARN [KUBERNETE] Re-init the cache on error error:[
ERROR REPORT:
Original Error: *trace.ConnectionProblemError watcher is closed: rpc error: code = Canceled desc = grpc: the client connection is closing
Stack Trace:
github.com/gravitational/teleport/lib/cache/cache.go:1173 github.com/gravitational/teleport/lib/cache.(*Cache).fetchAndWatch
github.com/gravitational/teleport/lib/cache/cache.go:971 github.com/gravitational/teleport/lib/cache.(*Cache).update
runtime/asm_amd64.s:1598 runtime.goexit
User Message: watcher is closed: rpc error: code = Canceled desc = grpc: the client connection is closing] cache/cache.go:977
2023-08-27T09:09:52Z INFO [UPLOAD:2]  Creating directory /var/lib/teleport/log/upload/corrupted. pid:6.2 service/service.go:2744
2023-08-27T09:09:52Z INFO [UPLOAD:2]  Creating directory /var/lib/teleport/log/upload/corrupted/default. pid:6.2 service/service.go:2744
2023-08-27T09:09:52Z INFO [UPLOAD:1]  File uploader is shutting down. pid:6.1 service/service.go:2783
2023-08-27T09:09:52Z INFO [UPLOAD:1]  File uploader has shut down. pid:6.1 service/service.go:2785
2023-08-27T09:09:52Z INFO [DIAG:1]    Shutting down gracefully. pid:6.1 service/service.go:3021
2023-08-27T09:09:52Z INFO [AUTH:1]    Shutting down immediately (auth service does not currently support graceful shutdown). pid:6.1 service/service.go:2032
2023-08-27T09:09:52Z INFO [PROC:COMP] upload completer will run every 5m0s events/complete.go:143
2023-08-27T09:09:52Z WARN [DIAG:1]    Diagnostic server exited with error: accept tcp [::]:3000: use of closed network connection. pid:6.1 service/service.go:3011
2023-08-27T09:09:52Z WARN [KUBERNETE] Heartbeat failed error reading from server: EOF. srv/heartbeat.go:282
2023-08-27T09:09:52Z INFO [AUTH:1]    Exited. pid:6.1 service/service.go:2047
2023-08-27T09:09:52Z INFO [PROC:2]    Detected Teleport component "kubernetes" is running in a degraded state. pid:6.2 service/state.go:106
2023-08-27T09:09:52Z WARN [AUTH:1]    TLS server exited with error: http: Server closed. pid:6.1 service/service.go:1922
2023-08-27T09:09:52Z INFO [UPLOAD:1]  File upload completer is shutting down. pid:6.1 service/service.go:2814
2023-08-27T09:09:52Z INFO [DIAG:1]    Exited. pid:6.1 service/service.go:3025
2023-08-27T09:09:52Z INFO [KUBERNETE] Exited. pid:6.1 service/kubernetes.go:300
2023-08-27T09:09:52Z WARN [KUBERNETE] Re-init the cache on error error:[
ERROR REPORT:
Original Error: *trace.ConnectionProblemError watcher is closed: error reading from server: EOF
Stack Trace:
github.com/gravitational/teleport/lib/cache/cache.go:1173 github.com/gravitational/teleport/lib/cache.(*Cache).fetchAndWatch
github.com/gravitational/teleport/lib/cache/cache.go:971 github.com/gravitational/teleport/lib/cache.(*Cache).update
runtime/asm_amd64.s:1598 runtime.goexit
User Message: watcher is closed: error reading from server: EOF] cache/cache.go:977
2023-08-27T09:09:52Z WARN [AUTH:1]    Mux serve failed. error:[
ERROR REPORT:
Original Error: *trace.ConnectionProblemError listener is closed
Stack Trace:
github.com/gravitational/teleport/lib/multiplexer/wrappers.go:137 github.com/gravitational/teleport/lib/multiplexer.(*Listener).Accept
crypto/tls/tls.go:66 crypto/tls.(*listener).Accept
github.com/gravitational/teleport/lib/multiplexer/tls.go:110 github.com/gravitational/teleport/lib/multiplexer.(*TLSListener).Serve
github.com/gravitational/teleport/lib/auth/middleware.go:268 github.com/gravitational/teleport/lib/auth.(*TLSServer).Serve.func1
runtime/asm_amd64.s:1598 runtime.goexit
User Message: listener is closed
listener is closed] auth/middleware.go:269
2023-08-27T09:09:52Z INFO [UPLOAD:1]  File upload completer has shut down. pid:6.1 service/service.go:2816
2023-08-27T09:09:52Z WARN [AUTH:1:CA] Re-init the cache on error error:[
ERROR REPORT:
Original Error: *trace.ConnectionProblemError watcher is closed: <nil>
Stack Trace:
github.com/gravitational/teleport/lib/cache/cache.go:1173 github.com/gravitational/teleport/lib/cache.(*Cache).fetchAndWatch
github.com/gravitational/teleport/lib/cache/cache.go:971 github.com/gravitational/teleport/lib/cache.(*Cache).update
runtime/asm_amd64.s:1598 runtime.goexit
User Message: watcher is closed: <nil>] cache/cache.go:977
2023-08-27T09:09:52Z WARN             Failed to get agent metadata: context canceled inventory/inventory.go:128
2023-08-27T09:09:52Z INFO             The old service was successfully shut down gracefully. service/service.go:777
2023-08-27T09:09:52Z WARN [KUBERNETE] Restart watch on error: watcher is closed: error reading from server: EOF. pid:6.2 resource-kind:lock services/watcher.go:246
2023-08-27T09:09:57Z INFO [PROC:2]    Teleport component "kubernetes" is recovering from a degraded state. pid:6.2 service/state.go:120
2023-08-27T09:10:12Z INFO [PROC:2]    Teleport component "kubernetes" has recovered from a degraded state. pid:6.2 service/state.go:124
```
teleprt-proxy pod:
```
2023-08-27T09:09:18Z INFO             domain teleport-auth-v12.teleport.svc.cluster.local not found common/wait.go:148
2023-08-27T09:09:18Z INFO             no endpoints found, exiting with success code common/wait.go:131
```

**Comment by hugoShaka** [comment]
Could you provide:
- The status of each teleport-cluster pod (`kubectl get pods -n <you namespace>`)
- The values.yaml you are using
- If you have enabled the teleport-operator

**Comment by hugoShaka** [comment]
@EinarArnason thank you for the bug report, we identified the issue. This is a consequence of disabling and enabling the operator in Helm. When disabling, Helm flags CRDs for deletion. However deletion is blocked until all CRs are deleted, which can't happen as the operator is not running.

When the operator is back healthy, it will delete all resources as instructed by Kubernetes.

This is a bug in our Helm chart, the chart should not delete CRDs when the user disables the operator.

**Comment by hugoShaka** [comment]
Fixed in Teleport 15.2, this will not be backported to previous major versions.
